### DISCUSSION TITLE
Working with multiple HL7 messages in one Json message

### DISCUSSION THREAD

**Discussion by rutledj** [nextgenhealthcare/connect#4889] [Q&A] [answered]
I have a listener channel that receives a JSON message which contains an HL7 formatted ack in the acknowledgement node as well as a HL7 formatted original message in the message_data mode.  

I can assign either of these to the msg object using msg = SerializerFactory.getSerializer('HL7V2').toXML(msg.acknowledgement); or 
msg = SerializerFactory.getSerializer('HL7V2').toXML(msg.message_data); 

I can then work with it as you would a normal msg.  What I want to do is to read both of these HL7 messages into two different variables:

var origMsg = new XML();

msg = SerializerFactory.getSerializer('HL7V2').toXML(msg.acknowledgement);
origMsg = SerializerFactory.getSerializer('HL7V2').toXML(msg.message_data);

At this point, orgMsg doesn't appear to contains any data: 
origMsg <?xml version="1.0" encoding="UTF-8"?><HL7Message><undefined></undefined></HL7Message>

The problem is, origMsg doesn't behave like the msg object. I can't access anything in the message using standard conventions. For example, I'd like to set the ack msg MSH-6.1 = the original message MSH-6.1. but I can't figure out how to do that.

(msg['MSH']['MSH.6']['MSH.6.1'] = origMsg['MSH']['MSH.6']['MSH.6.1'].toString()).  

Any suggestions?

**Comment by jonbartels** [comment]
`origMsg` is null because in the line before it you immediately changed the value of `msg` to an object which does not have the `message_data` property.

You need to juggle it around differently:
```
var msgAck = SerializerFactory.getSerializer('HL7V2').toXML(msg.acknowledgement);
var msgData = SerializerFactory.getSerializer('HL7V2').toXML(msg.message_data);
msg = msgAck;
```

**Comment by rutledj** [comment]
Thanks for the reply. Is there any way to have two variables that behave like the msg object? I.E. msg1 = msgAck, msg2 = msgData?
I want to be able to access data from both variables.

**Reply by pacmano1** [reply]
I don' think you want to "reset" msg here by assigning it.   Rather use two vars each that represent a message (msg remains your JSON stuff)

**Reply by pacmano1** [reply]
Can you post a sample JSON blob?

**Reply by pacmano1** [reply]
I was last to the party - I see @jonbartels posted the conversion.  That would typically fail if the field in JSON is not correct.  
 
See below for corrected conversion.

**Comment by jonbartels** [comment]
Try it!

The code I suggested should work the way you want.

```
var msgAck = SerializerFactory.getSerializer('HL7V2').toXML(msg.acknowledgement);
var msgData = SerializerFactory.getSerializer('HL7V2').toXML(msg.message_data);

logger.error('msgAck: ' + msgAck['MSH']['MSH.6']['MSH.6.1'].toString() )
logger.error('msgData: ' + msgData['MSH']['MSH.6']['MSH.6.1'].toString() )
```

If that does not work then I would look at the serializer call. I think it is correct though.

**Comment by rutledj** [comment]
Yep. Something with the conversion it doesn't like. I'll play around with it.

Wrapped com.mirth.connect.donkey.model.message.MessageSerializerException: Error converting ER7 to XML

**Comment by rutledj** [comment]
Here is the sample JSON I'm getting. I'm wondering if the '\r' in the message is throwing it off. On the other hand, it works fine when assigned to the msg object,

This looks ugly in preview, and it is putting a strikethrough in part of the message.  Not sure if there is a better way to post this.

{
  "public_health_id" : 3756982,
  "message_control_id" : "2909",
  "message_tracking_id" : "2909",
  "environment_id" : 1,
  "practice_id" : null,
  "provider_identifier" : "6418",
  "patient_identifier" : "E3932575",
  "acknowledgement" : "MSH|^~\\&|WebIZ.21.10.0.0|NMSIIS||RMAA6075|20211206155215-0700||ACK^V04^ACK|NMSIIS20211206521591|P|2.5.1|||NE|NE|||||Z23^CDCPHINVS\rMSA|AE|2904\rERR||PID^1^6^1^7|999^ApplicationError^HL70357|W|4^Invalid value^HL70533^IZ-66^XPN_M.7 (Name Type code) SHALL be valued \"M\"^L||NumericPath: PID[1].6[1].7, NamePath: PID/MothersMaidenName[0]/NameTypeCode, RuleId: 137, ApplicationErrorCode: IZ-66|PID-6 (MothersMaidenName[1]): Unrecognized/unsupported value: [].  Expectation: value in XPN.7 (Name Type code) SHALL be valued \"M\" for the field Mother's Maiden Name.\rERR||NK1^1^5^1^2|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: NK1[1].5[1].2, NamePath: NK1[0]/PhoneNumber[0]/TelecommunicationUseCode, RuleId: 1, ApplicationErrorCode: WEBIZ-701|NK1-5 (PhoneNumber[1]): Unrecognized/unsupported value [PRS].  Expectation: value must be one of the following: (is null or empty or exists in table HL70201).\rERR||NK1^1^5^1^2|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: NK1[1].5[1].2, NamePath: NK1[0]/PhoneNumber[0]/TelecommunicationUseCode, RuleId: 21, ApplicationErrorCode: WEBIZ-701|NK1-5 (PhoneNumber[1]): Unrecognized/unsupported value [PRS].  Expectation: value is in the list of supported telecommunication codes [ASN,BPN,NET,ORN,PRN,WPN] (telecommunications not in this list will not load).\rERR||ORC^1^20^1^3|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].20[1].3, NamePath: ORDER[0]/ORC/AdvancedBeneficiaryNoticeCode/NameOfCodingSystem, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-20 (AdvancedBeneficiaryNoticeCode): Unrecognized/unsupported value [LISA].  Expectation: value must be one of the following: (is null or empty or exists in table HL70396).\rERR||ORC^1^21^1^9|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].21[1].9, NamePath: ORDER[0]/ORC/OrderingFacilityName[0]/NameRepresentationCode, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-21 (OrderingFacilityName[1]): Unrecognized/unsupported value [RMANM MASTHEAD].  Expectation: value must be one of the following: (is null or empty or exists in table HL70465).\rERR||ORC^1^22^1^7|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].22[1].7, NamePath: ORDER[0]/ORC/OrderingFacilityAddress[0]/AddressType, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-22 (OrderingFacilityAddress[1]): Unrecognized/unsupported value [9987400].  Expectation: value must be one of the following: (is null or empty or exists in table HL70190).\rZSA|AW^Application Warning - One or more ERR segments returned with at most a severity of W for Warning^ENV0008|62359014^^^NMSIIS^HL7LogIdIncomming~6540381^^^NMSIIS^WebServiceLogIdIncomming~2904^^^RMAA6075^MessageControlId|2207882^^^NMSIIS^SR\r",
  "date_ack_sent" : null,
  "error_message" : null,
  "warnings" : null,
  "bridgeway_id" : 8832,
  "message_data" : "MSH|^~\\&||DAFTX SERVICE AREA|||20211208104007|6418|VXU^V04^VXU_V04|2909|P|2.5.1|||ER|AL|||||Z22^CDCPHINVS\rPID|1||E3932575^^^EPIC^MR||DOE^JANE^^^^^L|Summers|19510510|F||^Unknown^HL70005|999 Mickey Drive^^FORT WORTH^TX^76134||(682)2150566^PRN^PH^^^682^2150566|^PRN^PH^^^682^5555555|ENG|UNKNOWN||999665555||||2186-5^NOT HISPANIC OR LATINO^CDCREC||||||||N\rPD1|||DAFTX HUGULEY^^1089102|1982654554^JAGADISH^LALITHA^^^^^^NPI^L^^^NPI^^^^^^^^OTHER||||||||||||A|20210714\rORC|RE|100042542^EPC|8316202^Acumen 2.0||||^^^20211208^^^^||20211208102021|1245277490^SINGH^SANDEEP^^^^^^NPI^L^^||||||||||1245277490^SINGH^SANDEEP^^^^^^NPI^L^^^NPI^^^^^^^^M.D.|1089102100^^^1089102^^^^^DAFTX HUGULEY|^^^^^817^3365060||||CLISUP^EPIC SUPPORT^^10101102^EMH MED SURG|||||||||||O|PP NO COSIGN\rRXA|0|1|20211208|20211208|186^INFLUENZA, MDCK, QUADRIVALENT, WITH PRESERVATIVE^CVX|0.5|mL||00^New immunization record^NIP001|6418^GARCIA^NAOMI^^^^^^^L|1089102100^^^SIISCLIENT11595^^^^^DAFTX HUGULEY||||308523|20220630|SEQ^Seqirus^MVX|||CP|A|20211208102257\rRXR|IM^Intramuscular^HL70162|RD^Right Deltoid^HL70163\rOBX|1|CE|30956-7^Vaccine Type^LN|1|88^INFLUENZA, UNSPECIFIED^CVX||||||F|||20211208\rOBX|2|DT|29768-9^VIS Publication Date^LN|1|20190815||||||F|||20211208\rOBX|3|DT|29769-7^VIS Presentation Date^LN|1|20211208||||||F|||20211208\rOBX|4|CE|30963-3^Vaccine funding source^LN|2|VXC52^Public non-VFC^CDCPHINVS|1|||||F|||20211208\rOBX|5|CE|64994-7^Vaccine Funding Program Eligibility^LN|3|V01^Not VFC eligible^HL70064||||||F|||20211208|||VXC40^per immunization^CDCPHINVS\r",
  "transformed_data" : "MSH|^~\\&||1263470004|TXImmTrac|TxDSHS|20211208104201-0500|6418|VXU^V04^VXU_V04|3756982|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS||\rPID|1||E3932575^^^EPIC^MR||DOE^JANE^^^^^L|Summers|19510510|F||^Unknown^CDCREC|999 Mickey Drive^^FORT WORTH^TX^76134^^H||(682)2150566^PRN^PH^^^682^2150566|^PRN^PH^^^682^5555555|ENG|UNKNOWN||999665555||||2186-5^NOT HISPANIC OR LATINO^CDCREC||||||||N\rPD1|||DAFTX HUGULEY^^1089102|1982654554^JAGADISH^LALITHA^^^^^^NPI^L^^^NPI^^^^^^^^OTHER||||||||||||A|20210714\rORC|RE|100042542^EPC|8316202^Acumen 2.0||||||20211208102021|1245277490^SINGH^SANDEEP^^^^^^ACUMEN^L^^^PRN||||||||||1245277490^SINGH^SANDEEP^^^|1089102100^^^1089102^^^^^DAFTX HUGULEY|^^^^^817^3365060||||CLISUP^EPIC SUPPORT^^10101102^EMH MED SURG|||||\rRXA|0|1|20211208|20211208|186^INFLUENZA, MDCK, QUADRIVALENT, WITH PRESERVATIVE^CVX|0.5|mL^MilliLiter [SI Volume Units]^UCUM||00^New immunization record^NIP001|6418^GARCIA^NAOMI^^^^^^ACUMEN^L|1089102100^^^1263470004^^^^^DAFTX HUGULEY||||308523|20220630|SEQ^Seqirus^MVX|||CP|A|20211208102257\rRXR|C28161^Intramuscular^NCIT|RD^Right Deltoid^HL70163\rOBX|1|CE|30956-7^Vaccine Type^LN|1|88^INFLUENZA, UNSPECIFIED^CVX||||||F|||20211208\rOBX|2|DT|29768-9^VIS Publication Date^LN|1|20190815||||||F|||20211208\rOBX|3|DT|29769-7^VIS Presentation Date^LN|1|20211208||||||F|||20211208\rOBX|4|CE|30963-3^Vaccine funding source^LN|2|VXC52^Public non-VFC^CDCPHINVS||||||F|||20211208\rOBX|5|CE|64994-7^Vaccine Funding Program Eligibility^LN|3|V01^Not VFC eligible^HL70064||||||F|||20211208|||VXC40^per immunization^CDCPHINVS\r",
  "date_created" : "2021-12-08T15:40:16.743Z",
  "date_processed" : "2021-12-09T05:01:15.978Z",
  "status" : "processed"
}

**Reply by pacmano1** [reply]
Use code highlight

```
{
"public_health_id" : 3756982,
"message_control_id" : "2909",
"message_tracking_id" : "2909",
"environment_id" : 1,
"practice_id" : null,
"provider_identifier" : "6418",
"patient_identifier" : "E3932575",
"acknowledgement" : "MSH|^\&|WebIZ.21.10.0.0|NMSIIS||RMAA6075|20211206155215-0700||ACK^V04^ACK|NMSIIS20211206521591|P|2.5.1|||NE|NE|||||Z23^CDCPHINVS\rMSA|AE|2904\rERR||PID^1^6^1^7|999^ApplicationError^HL70357|W|4^Invalid value^HL70533^IZ-66^XPN_M.7 (Name Type code) SHALL be valued "M"^L||NumericPath: PID[1].6[1].7, NamePath: PID/MothersMaidenName[0]/NameTypeCode, RuleId: 137, ApplicationErrorCode: IZ-66|PID-6 (MothersMaidenName[1]): Unrecognized/unsupported value: []. Expectation: value in XPN.7 (Name Type code) SHALL be valued "M" for the field Mother's Maiden Name.\rERR||NK1^1^5^1^2|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: NK1[1].5[1].2, NamePath: NK1[0]/PhoneNumber[0]/TelecommunicationUseCode, RuleId: 1, ApplicationErrorCode: WEBIZ-701|NK1-5 (PhoneNumber[1]): Unrecognized/unsupported value [PRS]. Expectation: value must be one of the following: (is null or empty or exists in table HL70201).\rERR||NK1^1^5^1^2|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: NK1[1].5[1].2, NamePath: NK1[0]/PhoneNumber[0]/TelecommunicationUseCode, RuleId: 21, ApplicationErrorCode: WEBIZ-701|NK1-5 (PhoneNumber[1]): Unrecognized/unsupported value [PRS]. Expectation: value is in the list of supported telecommunication codes [ASN,BPN,NET,ORN,PRN,WPN] (telecommunications not in this list will not load).\rERR||ORC^1^20^1^3|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].20[1].3, NamePath: ORDER[0]/ORC/AdvancedBeneficiaryNoticeCode/NameOfCodingSystem, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-20 (AdvancedBeneficiaryNoticeCode): Unrecognized/unsupported value [LISA]. Expectation: value must be one of the following: (is null or empty or exists in table HL70396).\rERR||ORC^1^21^1^9|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].21[1].9, NamePath: ORDER[0]/ORC/OrderingFacilityName[0]/NameRepresentationCode, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-21 (OrderingFacilityName[1]): Unrecognized/unsupported value [RMANM MASTHEAD]. Expectation: value must be one of the following: (is null or empty or exists in table HL70465).\rERR||ORC^1^22^1^7|103^TableValueNotFound^HL70357|W|5^Table value not found^HL70533^WEBIZ-701^Table value not found^L||NumericPath: ORC[1].22[1].7, NamePath: ORDER[0]/ORC/OrderingFacilityAddress[0]/AddressType, RuleId: 1, ApplicationErrorCode: WEBIZ-701|ORC-22 (OrderingFacilityAddress[1]): Unrecognized/unsupported value [9987400]. Expectation: value must be one of the following: (is null or empty or exists in table HL70190).\rZSA|AW^Application Warning - One or more ERR segments returned with at most a severity of W for Warning^ENV0008|62359014^^^NMSIIS^HL7LogIdIncomming6540381^^^NMSIIS^WebServiceLogIdIncomming2904^^^RMAA6075^MessageControlId|2207882^^^NMSIIS^SR\r",
"date_ack_sent" : null,
"error_message" : null,
"warnings" : null,
"bridgeway_id" : 8832,
"message_data" : "MSH|^\&||DAFTX SERVICE AREA|||20211208104007|6418|VXU^V04^VXU_V04|2909|P|2.5.1|||ER|AL|||||Z22^CDCPHINVS\rPID|1||E3932575^^^EPIC^MR||DOE^JANE^^^^^L|Summers|19510510|F||^Unknown^HL70005|999 Mickey Drive^^FORT WORTH^TX^76134||(682)2150566^PRN^PH^^^682^2150566|^PRN^PH^^^682^5555555|ENG|UNKNOWN||999665555||||2186-5^NOT HISPANIC OR LATINO^CDCREC||||||||N\rPD1|||DAFTX HUGULEY^^1089102|1982654554^JAGADISH^LALITHA^^^^^^NPI^L^^^NPI^^^^^^^^OTHER||||||||||||A|20210714\rORC|RE|100042542^EPC|8316202^Acumen 2.0||||^^^20211208^^^^||20211208102021|1245277490^SINGH^SANDEEP^^^^^^NPI^L^^||||||||||1245277490^SINGH^SANDEEP^^^^^^NPI^L^^^NPI^^^^^^^^M.D.|1089102100^^^1089102^^^^^DAFTX HUGULEY|^^^^^817^3365060||||CLISUP^EPIC SUPPORT^^10101102^EMH MED SURG|||||||||||O|PP NO COSIGN\rRXA|0|1|20211208|20211208|186^INFLUENZA, MDCK, QUADRIVALENT, WITH PRESERVATIVE^CVX|0.5|mL||00^New immunization record^NIP001|6418^GARCIA^NAOMI^^^^^^^L|1089102100^^^SIISCLIENT11595^^^^^DAFTX HUGULEY||||308523|20220630|SEQ^Seqirus^MVX|||CP|A|20211208102257\rRXR|IM^Intramuscular^HL70162|RD^Right Deltoid^HL70163\rOBX|1|CE|30956-7^Vaccine Type^LN|1|88^INFLUENZA, UNSPECIFIED^CVX||||||F|||20211208\rOBX|2|DT|29768-9^VIS Publication Date^LN|1|20190815||||||F|||20211208\rOBX|3|DT|29769-7^VIS Presentation Date^LN|1|20211208||||||F|||20211208\rOBX|4|CE|30963-3^Vaccine funding source^LN|2|VXC52^Public non-VFC^CDCPHINVS|1|||||F|||20211208\rOBX|5|CE|64994-7^Vaccine Funding Program Eligibility^LN|3|V01^Not VFC eligible^HL70064||||||F|||20211208|||VXC40^per immunization^CDCPHINVS\r",
"transformed_data" : "MSH|^~\&||1263470004|TXImmTrac|TxDSHS|20211208104201-0500|6418|VXU^V04^VXU_V04|3756982|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS||\rPID|1||E3932575^^^EPIC^MR||DOE^JANE^^^^^L|Summers|19510510|F||^Unknown^CDCREC|999 Mickey Drive^^FORT WORTH^TX^76134^^H||(682)2150566^PRN^PH^^^682^2150566|^PRN^PH^^^682^5555555|ENG|UNKNOWN||999665555||||2186-5^NOT HISPANIC OR LATINO^CDCREC||||||||N\rPD1|||DAFTX HUGULEY^^1089102|1982654554^JAGADISH^LALITHA^^^^^^NPI^L^^^NPI^^^^^^^^OTHER||||||||||||A|20210714\rORC|RE|100042542^EPC|8316202^Acumen 2.0||||||20211208102021|1245277490^SINGH^SANDEEP^^^^^^ACUMEN^L^^^PRN||||||||||1245277490^SINGH^SANDEEP^^^|1089102100^^^1089102^^^^^DAFTX HUGULEY|^^^^^817^3365060||||CLISUP^EPIC SUPPORT^^10101102^EMH MED SURG|||||\rRXA|0|1|20211208|20211208|186^INFLUENZA, MDCK, QUADRIVALENT, WITH PRESERVATIVE^CVX|0.5|mL^MilliLiter [SI Volume Units]^UCUM||00^New immunization record^NIP001|6418^GARCIA^NAOMI^^^^^^ACUMEN^L|1089102100^^^1263470004^^^^^DAFTX HUGULEY||||308523|20220630|SEQ^Seqirus^MVX|||CP|A|20211208102257\rRXR|C28161^Intramuscular^NCIT|RD^Right Deltoid^HL70163\rOBX|1|CE|30956-7^Vaccine Type^LN|1|88^INFLUENZA, UNSPECIFIED^CVX||||||F|||20211208\rOBX|2|DT|29768-9^VIS Publication Date^LN|1|20190815||||||F|||20211208\rOBX|3|DT|29769-7^VIS Presentation Date^LN|1|20211208||||||F|||20211208\rOBX|4|CE|30963-3^Vaccine funding source^LN|2|VXC52^Public non-VFC^CDCPHINVS||||||F|||20211208\rOBX|5|CE|64994-7^Vaccine Funding Program Eligibility^LN|3|V01^Not VFC eligible^HL70064||||||F|||20211208|||VXC40^per immunization^CDCPHINVS\r",
"date_created" : "2021-12-08T15:40:16.743Z",
"date_processed" : "2021-12-09T05:01:15.978Z",
"status" : "processed"
}
```

**Reply by pacmano1** [reply]
https://github.com/nextgenhealthcare/connect/discussions/4824 and https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by pacmano1** [reply]
I think that JSON should have:
```
MSH|^\\&||.   // escape the backslash
```

**Comment by rutledj** [comment]
It actually is escaped but it appears to have been removed when I pasted it in here. See the screenshot.

![image](https://user-images.githubusercontent.com/6484700/145425827-b46384f0-6eae-493f-98d6-6751b5b82d51.png)

**Reply by pacmano1** [reply]
upload the JSON blob as an attachment.

**Comment by rutledj** [comment]
Here it is

[jsonblob.txt](https://github.com/nextgenhealthcare/connect/files/7686578/jsonblob.txt)

**Comment by pacmano1** [comment, marked as answer]
```JS
var serializationProperties = SerializerFactory.getDefaultSerializationProperties('HL7V2');

$c('msg1', new XML(SerializerFactory.getSerializer('HL7V2', serializationProperties, null).toXML(escape(msg['acknowledgement']))))
$c('msg2', new XML(SerializerFactory.getSerializer('HL7V2', serializationProperties, null).toXML(escape(msg['message_data']))))

$c('msg1_PID3', $c('msg1')['PID']['PID.3']['PID.3.1'].toString())
$c('msg2_PID3', $c('msg2')['PID']['PID.3']['PID.3.1'].toString())
```

**Comment by rutledj** [comment]
Interesting. So I tried that and then logging out the msg1

```
logger.error('msg1: ' + msg1.toString())
 ```


but it says msg1 is not defined.   How do you access it?

**Reply by pacmano1** [reply]
If you don't start using code highligthing, gonna stop helping.    You are logging out a local var.  `msg1` is a channelMap var.

**Reply by pacmano1** [reply]
And why log it out?  It is already viewable:

![image](https://user-images.githubusercontent.com/44065187/145438734-d5ee2f7c-1235-45ad-a062-9a906cd814c6.png)

**Comment by rutledj** [comment]
Gotcha. Thank you for your help.

**Reply by pacmano1** [reply]
Do mark something in the thread as an asnwer please.  The "mark as answer" button.

**Comment by rutledj** [comment]
Follow-up question if I may.  I'm getting a different error (different message) saying this:

```
org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
```
The incoming Json has been stripped down to basically nothing trying to remove the error. 

```
{
  "public_health_id" : 3756982,
  "message_control_id" : "2909",
  "message_tracking_id" : "2909",
  "environment_id" : 1,
  "practice_id" : null,
  "provider_identifier" : "6418",
  "patient_identifier" : "E3932575",
  "acknowledgement" : null,
  "date_ack_sent" : null,
  "error_message" : null,
  "warnings" : null,
  "bridgeway_id" : 8832,
  "message_data" : null,
  "transformed_data" : null,
  "date_created" : "2021-12-08T15:40:16.743Z",
  "date_processed" : "2021-12-09T05:01:15.978Z",
  "status" : "processed"
}
```

Any ideas on this?  I'm thinking it may have to do with the conversion between JSON and HL7 in the mirth source since I'm setting  inbound to JSON and outbound to HL7v2.x in the Set Data Types box.

**Reply by jonbartels** [reply]
`org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.` always comes from parsing XML that is null, empty, not-XML, or which has hidden characters/whitespace before the XML starts.

You have a JSON message and are getting an XML error. Are the channel datatypes set correctly?

**Comment by rutledj** [comment]
The source Inbound is set to JSON and the remaining are set to HL7v2.x

**Reply by pacmano1** [reply]
What I wrote above does not dispatch two hl7 messages to destinations.   Nor does it change the original message so JSON would be expected for all data types.  So, please describe your exact use case - probably create a new issue since the topic seems to be changing :).

**Comment by rutledj** [comment]
Yes. I think that was the issue. I was originally setting the msg object to the acknowledgment HL7 from the JSON. With your code, that was removed, so no HL7 was getting sent to the destinations. I think it is working as I need it now. Thanks much for the help.
